documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+# Summary
+A new API for generic maths in Silk.NET.Maths 3.X, leveraging modern .NET features such as `INumber<T>` and vectorization.
+
+This API aims to replace the existing implementation of Silk.NET.Maths.
+
+# Contributors
+- Maxine H (@uwx)
+
+# Current Status
+- [x] Proposed
+- [ ] Discussed with API Review Board (ARB)
+- [ ] Approved
+- [ ] Implemented
+
+# Design Decisions
+- Should replace the existing APIs without too much growing pains
+- Should try to roughly match System.Numerics APIs, with concessions for design oversights in those APIs
+- Should use vectorized implementations for methods when possible
+- Prefer generic specialization and extension methods, and keep the type of `T` as wide as possible in the struct itself
+- Primitives are read-only, `with` preferred for modification
+- Public fields instead of properties
```

Comment:
I would strongly recommend against this. It is one of the worst design decisions for the types in System.Numerics and has made it extremely difficult to do new optimizations and code cleanup.

Touching fields directly is also an anti-pattern towards SIMD and negatively impacts the JIT's ability to optimize.

Reply:
Noted. What about `Vector2D<T>.Components` (or AsSpan as you suggested) - will that scenario behave as expect with auto-properties?

Reply:
Yes (at least for unmanaged T). Structs in C# are sequential by default and backing fields for auto properties are emitted in declaration order.

For AsSpan you’d then be creating it from ref this, and reinterpret casting that from Vector2D<T> to T, and so it’s probably better to define the fields as private and not use auto props.

The general problem with public fields is you can never change the implementation later. Imagine we had an identical type in the BCL eventually and then you can’t change the implementation to be a single field of that type. — Matrix4x4, for example, would have been better implemented as 4x Vector4 fields than 16x float fields, for example. Or maybe even as Vector128<float>, etc. but we can’t change it today and are stuck doing really bad hacks 

Or consider if we get some specialization support in the future, where the fields could differ based on the T. You’d then want to change the backing fields in that case for primitive types, presumably

Reply:
Public fields are never a good idea (outside interop where it can’t change because they are functionally public in C already)

And it’s always better (and no worse in efficiency) to just expose it as public properties